File: mobile-client/android/UseCase/Product/Detail/LoadProductDetailUseCase.kt
package app.mobiling.client.client.usecase.product.detail

import app.mobiling.client.client.contract.product.detail.ProductDetailPayload
import app.mobiling.client.client.data.product.detail.ProductDetailGateway

/**
 * Marketing America Corp. Oleksandr Tishchenko
 */
class LoadProductDetailUseCase(
    private val gateway: ProductDetailGateway,
) {
    suspend operator fun invoke(productId: String): ProductDetailPayload? = gateway.loadProductDetail(productId)
}

File: mobile-client/ios/UseCase/Product/Detail/LoadProductDetailUseCase.swift
import Foundation

// Marketing America Corp. Oleksandr Tishchenko
public struct LoadProductDetailUseCase: Sendable {
    private let gateway: any ProductDetailGateway

    public init(gateway: any ProductDetailGateway) {
        self.gateway = gateway
    }

    public func callAsFunction(productId: String) async throws -> ProductDetailPayload? {
        try await gateway.loadProductDetail(productId: productId)
    }
}
